File: apps/frontend/src/adapters/web/captures.ts
import { invoke } from "./core";
import type { CaptureInput, CaptureReceipt } from "@/features/quick-add/types";
export const submitCapture = (input: CaptureInput) =>
  invoke<CaptureReceipt>("submit_capture", { input });
export const getCapture = (id: string) => invoke<CaptureReceipt>("get_capture", { id });
export const getCaptureReviews = (page = 0, pageSize = 25, reason?: string) =>
  invoke<CaptureReceipt[]>("get_capture_reviews", { page, pageSize, reason });
export const resolveCaptureReview = (
  id: string,
  resolution: import("@/features/quick-add/types").CaptureResolution,
) => invoke<CaptureReceipt>("resolve_capture_review", { id, resolution });
export const dismissCaptureReview = (id: string, version: number) =>
  invoke<CaptureReceipt>("dismiss_capture_review", { id, version });
export const getCaptureSettings = () =>
  invoke<import("@/features/quick-add/types").CaptureSettings>("get_capture_settings");
export const updateCaptureSettings = (
  settings: import("@/features/quick-add/types").CaptureSettings,
) => invoke<void>("update_capture_settings", { settings });
export const listCaptureTokens = () =>
  invoke<import("@/features/quick-add/types").CaptureToken[]>("list_capture_tokens");
export const createCaptureToken = (name: string, expiresAt: string | null) =>
  invoke<import("@/features/quick-add/types").CaptureToken & { token: string }>(
    "create_capture_token",
    { name, expiresAt },
  );
export const revokeCaptureToken = (id: string) => invoke<void>("revoke_capture_token", { id });

export const retryCapture = (id: string, version: number) =>
  invoke<CaptureReceipt>("retry_capture", { id, version });
export const clearCaptureSource = (id: string, version: number) =>
  invoke<CaptureReceipt>("clear_capture_source", { id, version });

export const getCaptureUsage = () =>
  invoke<import("@/features/quick-add/types").CaptureUsage>("get_capture_usage");

File: apps/tauri/src/commands/captures.rs
use crate::context::ServiceContext;
use std::sync::Arc;
use tauri::State;
use wealthfolio_core::captures::{Capture, CaptureInput};
#[tauri::command]
pub async fn submit_capture(
    state: State<'_, Arc<ServiceContext>>,
    input: CaptureInput,
) -> Result<Capture, String> {
    state
        .capture_service
        .accept("application", input)
        .await
        .map_err(|e| e.to_string())
}
#[tauri::command]
pub async fn get_capture(
    state: State<'_, Arc<ServiceContext>>,
    id: String,
) -> Result<Capture, String> {
    state
        .capture_service
        .get(&id, None)
        .map_err(|e| e.to_string())?
        .ok_or_else(|| "Capture not found".into())
}

#[tauri::command]
pub async fn get_capture_reviews(
    state: State<'_, Arc<ServiceContext>>,
    page: Option<u32>,
    page_size: Option<u32>,
    reason: Option<String>,
) -> Result<Vec<Capture>, String> {
    state
        .capture_service
        .reviews(
            page.unwrap_or(0),
            page_size.unwrap_or(25),
            reason.as_deref(),
        )
        .map_err(|e| e.to_string())
}
#[tauri::command]
pub async fn resolve_capture_review(
    state: State<'_, Arc<ServiceContext>>,
    id: String,
    resolution: wealthfolio_core::captures::ResolveReview,
) -> Result<Capture, String> {
    state
        .capture_service
        .resolve(&id, resolution)
        .await
        .map_err(|e| e.to_string())
}
#[tauri::command]
pub async fn dismiss_capture_review(
    state: State<'_, Arc<ServiceContext>>,
    id: String,
    version: i64,
) -> Result<Capture, String> {
    state
        .capture_service
        .dismiss(&id, version)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_capture_settings(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<wealthfolio_core::captures::settings::CaptureSettings, String> {
    state.capture_service.settings().map_err(|e| e.to_string())
}
#[tauri::command]
pub async fn update_capture_settings(
    state: State<'_, Arc<ServiceContext>>,
    settings: wealthfolio_core::captures::settings::CaptureSettings,
) -> Result<(), String> {
    state
        .capture_service
        .configure(settings)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn retry_capture(
    state: State<'_, Arc<ServiceContext>>,
    id: String,
    version: i64,
) -> Result<Capture, String> {
    state
        .capture_service
        .retry(&id, version)
        .await
        .map_err(|e| e.to_string())
}
#[tauri::command]
pub async fn clear_capture_source(
    state: State<'_, Arc<ServiceContext>>,
    id: String,
    version: i64,
) -> Result<Capture, String> {
    state
        .capture_service
        .clear_source(&id, version)
        .await
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_capture_usage(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<wealthfolio_core::captures::settings::CaptureUsage, String> {
    state.capture_service.usage().map_err(|e| e.to_string())
}
